core/src/main/java/org/infinispan/CacheStreamable.java

```diff
@@ -0,0 +1,120 @@
+package org.infinispan;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.BaseStream;
+
+/**
+ * Interface that signifies which streams are able to be used with a {@link Cache}.  This method
+ * is useful to hold a reference to any of the types while still being able to invoke some methods.
+ */
+public interface CacheStreamable<T, S extends BaseStream<T, S>> extends BaseStream<T, S> {
+   /**
+    * This would disable sending requests to all other remote nodes compared to one at a time. This can reduce memory
+    * pressure on the originator node at the cost of performance.
+    * <p>Parallel distribution is enabled by default except for {@link CacheStream#iterator()} &
+    * {@link CacheStream#spliterator()}</p>
+    * @return a stream with parallel distribution disabled
+    */
+   CacheStreamable sequentialDistribution();
+
+   /**
+    * This would enable sending requests to all other remote nodes when a terminal operator is performed.  This
+    * requires additional overhead as it must process results concurrently from various nodes, but should perform
+    * faster in the majority of cases.
+    * <p>Parallel distribution is enabled by default except for {@link CacheStream#iterator()} &
+    * {@link CacheStream#spliterator()}</p>
+    * @return a stream with parallel distribution enabled.
+    */
+   CacheStreamable parallelDistribution();
+
+   /**
+    * Filters which entries are returned by what segment they are present in.  This method can be substantially more
+    * efficient than using a regular {@link CacheStream#filter(Predicate)} method as this can control what nodes are
+    * asked for data and what entries are read from the underlying CacheStore if present.
+    * @param segments The segments to use for this stream operation.  Any segments not in this set will be ignored.
+    * @return a stream with the segments filtered.
+    */
+   CacheStreamable filterKeySegments(Set<Integer> segments);
+
+   /**
+    * Filters which entries are returned by only returning ones that map to the given key.  This method will <b>always</b>
+    * be faster than a regular {@link CacheStream#filter(Predicate)} if any keys must be retrieved remotely or if a
```

Comment:
I wouldn't say *always*, `cache.entrySet().stream().filterKeys(cache.keySet())` is very likely to be slower regardless of clustering/persistence :)